lib/web/purchases_flutter_web.dart

```diff
@@ -0,0 +1,440 @@
+        code: _unknownErrorCode,
+        message: 'Error calling static method $methodName: $e',
+      );
+    }
+  }
+
+  JSObject _callInstanceMethod(
+      String methodName,
+      List<dynamic> args,
+  ) {
+    final purchases = _getInstance();
+    final jsArgs = _processArgs(args);
+    return purchases.callMethodVarArgs(methodName.toJS, jsArgs);
+  }
+
+  JSObject _getInstance() {
+    final purchases = _getStaticPurchasesCommon();
+
+    if (!_isConfigured()) {
+      throw PlatformException(
+        code: _configurationErrorCode,
+        message: 'Purchases SDK not configured. Call configure first.',
+      );
+    }
+
+    return purchases.callMethod('getInstance'.toJS);
+  }
+
+  JSObject _getStaticPurchasesCommon() {
+    final purchasesHybridMappings = globalContext['PurchasesHybridMappings'] as JSObject;
+    JSObject? purchasesCommon;
+    if (purchasesHybridMappings.has('PurchasesCommon')) {
+      purchasesCommon = purchasesHybridMappings['PurchasesCommon'] as JSObject;
+    }
+    if (purchasesCommon == null) {
+      throw PlatformException(
+        code: _configurationErrorCode,
+        message: 'Purchases SDK not found on window object after loading.',
+      );
+    }
+    return purchasesCommon;
+  }
+
+  Future _callStaticMethodReturningPromise(
+      String methodName,
+      List<dynamic> args,
+      ) async {
+    final promise = _callStaticMethod(methodName, args) as JSPromise;
+    return promise.toDart
+        .catchError((error) => throw _processError(error));
+  }
+
+  Future<Map<String, dynamic>> _getMapFromInstanceMethod(
+      String methodName,
+      List<dynamic> args,
+  ) async {
+    final promise = _callInstanceMethod(methodName, args) as JSPromise;
+
+    return promise.toDart
+        .then(_convertJsRecordToMap)
+        .catchError((error) => throw _processError(error));
+  }
+
+  Future<Map<String, dynamic>?> _getNullableMapFromInstanceMethod(
+      String methodName,
+      List<dynamic> args,
+  ) async {
+    final promise = _callInstanceMethod(methodName, args) as JSPromise;
+
+    return promise.toDart
+        .then((value) {
+          if (value == null) {
+            return null;
+          }
+          return _convertJsRecordToMap(value);
+        })
+        .catchError((error) => throw _processError(error));
+  }
+
+  List<JSAny?> _processArgs(List<dynamic> args) {
+    final jsArgs = <JSAny?>[];
+    for (final arg in args) {
+      if (arg is Map<String, dynamic>) {
+        jsArgs.add(arg.jsify());
+      } else if (arg is List) {
+        jsArgs.add(arg.jsify());
+      } else if (arg is String) {
+        jsArgs.add(arg.toJS);
+      } else if (arg is int) {
+        jsArgs.add(arg.toJS);
+      } else if (arg is bool) {
+        jsArgs.add(arg.toJS);
+      } else if (arg == null) {
+        jsArgs.add(null);
+      } else {
+        throw ArgumentError(
+          'Unsupported argument type: ${arg.runtimeType}',
+        );
+      }
+    }
+    return jsArgs;
+  }
```

Comment:
Dart 3 also has support for pattern matching, e.g.:

```suggestion
  List<JSAny?> _processArgs(List<dynamic> args) {
    return <JSAny?>[
      for (final arg in args)
        switch (arg) {
          final Map<String, dynamic> arg => arg.jsify(),
          final List arg => arg.jsify(),
          final String arg => arg.toJS,
          final int arg => arg.toJS,
          final bool arg => arg.toJS,
          null => null,
          _ => throw ArgumentError(
              'Unsupported argument type: ${arg.runtimeType}',
            ),
        },
    ];
  }
```